--- /usr/src/linux-3.10.1-guest/arch/x86/kvm/x86.c	2013-08-29 10:32:58.014049861 -0700
+++ x86.c	2013-08-29 18:54:28.335334406 -0700
@@ -63,6 +63,7 @@
 #include <asm/xcr.h>
 #include <asm/pvclock.h>
 #include <asm/div64.h>
+#include <asm/vmdd.h>
 
 #define MAX_IO_MSRS 256
 #define KVM_MAX_MCE_BANKS 32
@@ -5481,6 +5482,18 @@
 	case KVM_HC_VAPIC_POLL_IRQ:
 		ret = 0;
 		break;
+	case KVM_HYPERCALL_CONNECT_TO_TCSERVICE://13579
+		ret = vmdd_connect(vcpu);
+		break;
+	case KVM_HYPERCALL_DISCONNECT_FROM_TCSERVICE: //13580
+		ret = vmdd_disconnect(vcpu);
+		break;
+	case KVM_HYPERCALL_READ_FROM_TCSERVICE://13581
+		ret = vmdd_read(vcpu, (struct file *) a0, (char *)a1, (ssize_t) a2, (loff_t *) a3);
+		break;
+	case KVM_HYPERCALL_WRITE_TO_TCSERVICE:	//13580
+		ret = vmdd_write(vcpu, (struct file *) a0, (char *)a1, (ssize_t) a2, (loff_t *) a3);
+		break;
 	default:
 		ret = -KVM_ENOSYS;
 		break;
